sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +80,130 @@ export class NotificationService extends RecordDataModelService<Notification> {
+   * @param permanentFailureError Error details of a permanent failure if occurred while processing the notification.
    * @returns result of the record updated.
    */
   async updateNotification(
     notificationId: number,
-    entityManager?: EntityManager,
+    permanentFailureError?: PermanentFailureError[],
   ): Promise<UpdateResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    return notificationRepo.update(
+    return this.repo.update(
       {
         id: notificationId,
       },
-      { dateSent: new Date() },
+      { dateSent: new Date(), permanentFailureError },
     );
   }
 
   /**
-   * Invokes the sendEmailNotification method of the GC Notification service.
-   * @param notificationId notification id used to retrieve the payload that will be passed
-   * as a parameter to the sendMailNotification method.
-   * @param entityManager optional repository that can be provided, for instance,
-   * to execute the command as part of an existing transaction. If not provided
-   * the local repository will be used instead.
-   * @returns GC Notify API call response.
+   * Process all the unsent notifications with a polling limit.
+   * Processing continues recursively until all the records are processed.
+   * Call GCNotify to send email notification.
+   * @param pollingLimit
+   * @param notificationsProcessed optional param used to
+   * increment count on recursion.
+   * @param notificationsSuccessfullyProcessed optional param used to
+   * increment count on recursion.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
-      select: {
-        id: true,
-        messagePayload: true,
-      },
-      where: {
-        id: notificationId,
-      },
-    });
-    // Call GC Notify send email method.
-    const gcNotifyResult = await this.gcNotifyService.sendEmailNotification(
-      notification.messagePayload,
-    );
+  async processUnsentNotifications(
+    pollingLimit: number,
+    notificationsProcessed = 0,
+    notificationsSuccessfullyProcessed = 0,
+  ): Promise<NotificationProcessingStatus> {
+    const notificationsToProcess = await this.repo
+      .createQueryBuilder("notification")
+      .select(["notification.id", "notification.messagePayload"])
+      .where("notification.dateSent IS NULL")
+      .limit(pollingLimit)
+      .getMany();
+
+    if (notificationsToProcess.length) {
+      this.logger.log(
+        `Processing ${notificationsToProcess.length} notifications`,
+      );
+
+      // Used to limit the number of asynchronous operations
+      // that will start at the same time.
+      const maxPromisesAllowed = os.cpus().length;
```

Comment:
Because we have the `pollingLimit` I was not expecting to have this as well.
As it is already done I am ok with it, if it was not, the pooling would be enough, IMO.
Not part of this PR but we used this technique so many times that it is passed the time to build a helper method to reuse this logic 😉